src/gadget/monotone.rs

```diff
@@ -0,0 +1,259 @@
+    pub fn configure(
+        meta: &mut ConstraintSystem<F>,
+        q_enable: impl FnOnce(&mut VirtualCells<'_, F>) -> Expression<F>,
+        value: Column<Advice>,
+    ) -> MonotoneConfig {
+        let range_table = meta.fixed_column();
+
+        let config = MonotoneConfig { range_table };
+
+        meta.lookup(|meta| {
+            let q_enable = q_enable(meta);
+            let range_table = meta.query_fixed(config.range_table, Rotation::cur());
+            let value_diff = {
+                let value_cur = meta.query_advice(value, Rotation::cur());
+                let value_prev = meta.query_advice(value, Rotation::prev());
+                if INCR {
+                    value_cur - value_prev
+                } else {
+                    value_prev - value_cur
+                }
+            };
+
+            // If strict monotone, we subtract diff by one
+            // to make sure zero lookup fail
+            let min_diff = Expression::Constant(F::from_u64(STRICT as u64));
+
+            vec![(q_enable * (value_diff - min_diff), range_table)]
+        });
+
+        config
+    }
+
+    pub fn load(&self, layouter: &mut impl Layouter<F>) -> Result<(), Error> {
+        layouter.assign_region(
+            || "range_table",
+            |mut meta| {
+                let max = RANGE - STRICT as usize;
+
+                for idx in 0..=max {
+                    meta.assign_fixed(
+                        || "range_table_value",
+                        self.config.range_table,
+                        idx,
+                        || Ok(F::from_u64(idx as u64)),
+                    )?;
+                }
+
+                Ok(())
+            },
+        )
+    }
+
+    pub fn construct(config: MonotoneConfig) -> Self {
+        Self {
+            config,
+            _marker: PhantomData,
+        }
+    }
+}
+
+impl<F: FieldExt, const RANGE: usize, const INCR: bool, const STRICT: bool> Chip<F>
+    for MonotoneChip<F, RANGE, INCR, STRICT>
+{
+    type Config = MonotoneConfig;
+    type Loaded = ();
+
+    fn config(&self) -> &Self::Config {
+        &self.config
+    }
+
+    fn loaded(&self) -> &Self::Loaded {
+        &()
+    }
+}
+
+#[cfg(test)]
+mod test {
+    use super::{MonotoneConfig, NonStrictMonoIncrChip, StrictMonoIncrChip};
+    use halo2::{
+        arithmetic::FieldExt,
+        circuit::{layouter::SingleChipLayouter, Layouter},
+        dev::{
+            MockProver,
+            VerifyFailure::{self, Lookup},
+        },
+        plonk::{Advice, Assignment, Circuit, Column, ConstraintSystem, Error, Selector},
+    };
+    use pasta_curves::pallas::Base;
+    use std::marker::PhantomData;
+
+    #[test]
+    fn mono_incr() {
+        #[derive(Clone, Debug)]
+        struct TestCircuitConfig {
+            q_enable: Selector,
+            value: Column<Advice>,
+            strict_mono_incr: MonotoneConfig,
+            non_strict_mono_incr: MonotoneConfig,
+        }
```

Comment:
I suggest separating the strict and non-strict tests, so we can identify which is expected to fail in the unit tests.

Reply:
Fixed in third commit